crates/ark/src/variables/methods.rs

```diff
@@ -0,0 +1,158 @@
+use harp::exec::RFunction;
+use harp::exec::RFunctionExt;
+use harp::r_null;
+use harp::r_symbol;
+use harp::utils::r_is_object;
+use harp::RObject;
+use libr::Rf_lang3;
+use libr::SEXP;
+use stdext::result::ResultOrLog;
+use strum::IntoEnumIterator;
+use strum_macros::Display;
+use strum_macros::EnumIter;
+use strum_macros::EnumString;
+use strum_macros::IntoStaticStr;
+
+use crate::modules::ARK_ENVS;
+
+#[derive(Debug, PartialEq, EnumString, EnumIter, IntoStaticStr, Display, Eq, Hash, Clone)]
+pub enum ArkGenerics {
+    #[strum(serialize = "ark_variable_display_value")]
+    VariableDisplayValue,
+
+    #[strum(serialize = "ark_variable_display_type")]
+    VariableDisplayType,
+
+    #[strum(serialize = "ark_variable_has_children")]
+    VariableHasChildren,
+
+    #[strum(serialize = "ark_variable_kind")]
+    VariableKind,
+}
+
+impl ArkGenerics {
+    // Dispatches the method on `x`
+    // Returns
+    //   - `None` if no method was found,
+    //   - `Err` if method was found and errored
+    //   - T, if method was found and was succesfully executed
+    pub fn try_dispatch<T>(
+        &self,
+        x: SEXP,
+        args: Vec<(String, RObject)>,
+    ) -> anyhow::Result<Option<T>>
+    where
+        // Making this a generic allows us to handle the conversion to the expected output
+        // type within the dispatch, which is much more ergonomic.
+        T: TryFrom<RObject>,
+        <T as TryFrom<harp::RObject>>::Error: std::fmt::Debug,
+    {
+        if !r_is_object(x) {
+            return Ok(None);
+        }
+
+        let generic: &str = self.into();
+        let mut call = RFunction::new("", "call_ark_method");
+
+        call.add(generic);
+        call.add(x);
+
+        for (name, value) in args.into_iter() {
+            call.param(name.as_str(), value);
+        }
+
+        let result = call.call_in(ARK_ENVS.positron_ns)?;
+
+        // No method for that object
+        if result.sexp == r_null() {
+            return Ok(None);
+        }
+
+        // Convert the result to the expected return type
+        match result.try_into() {
+            Ok(value) => Ok(Some(value)),
+            Err(err) => Err(anyhow!("Conversion failed: {err:?}")),
+        }
+    }
+
+    pub fn register_method(generic: Self, class: &str, method: RObject) -> anyhow::Result<()> {
+        let generic_name: &str = generic.into();
+        RFunction::new("", ".ps.register_ark_method")
+            .add(RObject::try_from(generic_name)?)
+            .add(RObject::try_from(class)?)
+            .add(method)
+            .call_in(ARK_ENVS.positron_ns)?;
+        Ok(())
+    }
+
+    pub fn register_method_from_package(
+        generic: Self,
+        class: &str,
+        package: &str,
+    ) -> anyhow::Result<()> {
+        let method = RObject::from(unsafe {
+            Rf_lang3(
+                r_symbol!(":::"),
+                r_symbol!(package),
+                r_symbol!(format!("{generic}.{class}")),
+            )
```

Comment:
IIRC, the motivation for storing a call `pkgname:::methodname` instead of the closure object was to ensure that:

1.  Everything worked seamlessly with `devtools::load_all()` and similar functions (no risk of calling an outdated method from a stale cache);
2.  An escape hatch remained available for monkey-patching a method in the package namespace (e.g., to enable workarounds for misbehaving methods).

Reply:
I updated to use `RCall`  in https://github.com/posit-dev/ark/pull/560/commits/aadef2b97c1f0c16fb528a5c13a0ac90e6b529a7